### PR TITLE
Revising equality check in Core to emit a guard

### PR DESCRIPTION
**Pull request by nikswamy**
In case both sides are equational

With @meganfrisella, we observed today that if you try to prove `v.f == u.f` using Core, this results in an SMT guard for `v==u`, since it matches on the head symbol for the projector of `f` and then tries to prove the arguments equal. Of course, this doesn't work when `v` and `u` differ, although they agree on their `f` field.

The patch here tries to mimic the behavior of Rel. If the head symbol of both sides of an equation `t1 == t2` are equatable (as decided by Rel, i.e., they are marked equational), and if emitting a guard is ok, then Core will prefer to emit a guard rather than unfolding and trying the prove the equation structurally.

### CODE REVIEW

File: src/typechecker/FStar.TypeChecker.Core.fst [hunk 1]
```diff
@@ -1002,10 +1005,21 @@ let rec check_relation (g:env) (rel:relation) (t0 t1:typ)
...
-            (check_relation g EQUALITY head0 head1 ;!
-             check_relation_args g EQUALITY args0 args1)
-            (fun _ -> maybe_unfold_side_and_retry Both t0 t1)
+          (* If we're proving equality and SMT queries are ok, 
+             defer things like `v.v1 == u.v1` to SMT
+             rather than matching on the v1 projector and
+             trying to prove `v == u` *)
+          if guard_ok &&
+            (rel=EQUALITY) && 
+            equatable g t0 &&
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.Core.fst:1014]: Is the correct check `equatable g t0 && equatable g t1` or `equatable g t0 || equatable g t1` ... we do the latter at another place in this function.

Comment (nikswamy) [reply src/typechecker/FStar.TypeChecker.Core.fst]: I've revised this

File: src/typechecker/FStar.TypeChecker.Core.fst [hunk 2]
```diff
@@ -1002,10 +1005,28 @@ let rec check_relation (g:env) (rel:relation) (t0 t1:typ)
...
+          in
+          if guard_ok &&
+            (rel=EQUALITY) && 
+            (equatable g t0 || equatable g t1)
+          then (
+            handle_with 
+              (no_guard (compare_head_and_args ()))
+              (fun _ -> emit_guard t0 t1)
+          )
+          else compare_head_and_args ()
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Core.fst:1029]: Thanks to discussion with @mtzguido, this check is now revised as per the comment.

If we're solving equality problems and *either* side is equatable, then first try structurally but *without* SMT. And if that fails, then if either fallback to SMT on the original pair of terms.

**Comment by nikswamy** [comment]
I just got an Everest green with this too.

**Pull request merged**